Zlibrary module and util for issue #170

vikas-kundu opened the pull request:
Zlibrary module and util for issue #170. Here is a gif of the same: https://asciinema.org/a/CE7i2Z6RnQGJKOQAYbSnkEBZ6

vikas-kundu committed:
Zlibrary module and util for issue #170

saeeddhqan commented on core/util/zlibrary.py:
```diff
@@ -0,0 +1,95 @@
+
+class main:
+
+	def __init__(self, query, exact, start_year, end_year, lang, extension, count):
```
great options

saeeddhqan commented on core/util/zlibrary.py:
```diff
@@ -0,0 +1,95 @@
+		self.lang = lang
+		self.extension = extension
+		self.count = count
+		self.limit = math.ceil(count/50) #No. of pages to scrap. 50 entries per page so total = ceil of limit/50
```
thanks for the comments. always put a space after # No...

saeeddhqan commented on core/util/zlibrary.py:
```diff
@@ -0,0 +1,95 @@
+		self.limit = math.ceil(count/50) #No. of pages to scrap. 50 entries per page so total = ceil of limit/50
+		self._books_data = ''
+		self._articles_data = ''
+		self.framework._global_options['rand_agent'] = True
```
after your work. take the rand_agent place to the previous value.

saeeddhqan commented on core/util/zlibrary.py:
```diff
@@ -0,0 +1,95 @@
+			try:
+				scrap_result += self.framework.request(f'{url}{payload}&page={page_no}' , cookies=cookies).text
+			except Exception as e:
+				self.framework.error(f'Zlibrary {source}, page {page_no} is missed!')
```
self.framework.error(f"...", 'util/zlibrary', 'search_all')

saeeddhqan commented on core/util/zlibrary.py:
```diff
@@ -0,0 +1,95 @@
+			total_entries += 50 #50 table entries present per page
+
+			try:
+				scrap_result += self.framework.request(f'{url}{payload}&page={page_no}' , cookies=cookies).text
```
f"..."

saeeddhqan commented on core/util/zlibrary.py:
```diff
@@ -0,0 +1,95 @@
+		try:
+			response = self.framework.request(welcome_page_url)
+		except Exception as e:
+			self.framework.error('Zlibrary is missed!')
```
self.framework.error(f"...", 'util/zlibrary', 'search')

saeeddhqan commented on core/util/zlibrary.py:
```diff
@@ -0,0 +1,95 @@
+			self.framework.error('Zlibrary is missed!')
+			return False
+		else:	
+			parsed_html = BeautifulSoup(response.text,features="lxml")
```
space and quote error

saeeddhqan commented on core/util/zlibrary.py:
```diff
@@ -0,0 +1,95 @@
+		else:	
+			parsed_html = BeautifulSoup(response.text,features="lxml")
+
+			zlib_books_url = parsed_html.body.find('span', attrs={'class':'domain-check-domain','data-mode':'books'}).text
```
space error

saeeddhqan commented on core/util/zlibrary.py:
```diff
@@ -0,0 +1,95 @@
+			parsed_html = BeautifulSoup(response.text,features="lxml")
+
+			zlib_books_url = parsed_html.body.find('span', attrs={'class':'domain-check-domain','data-mode':'books'}).text
+			zlib_articles_url = parsed_html.body.find('span', attrs={'class':'domain-check-domain','data-mode':'articles'}).text
```
space error

saeeddhqan commented on modules/search/zlibrary.py:
```diff
@@ -0,0 +1,105 @@
+	'sources': ('Zlibrary',),
+	'options': (
+		('query', '', True, 'Query string to search for', '-q', 'store', str),
+		('exact', False, False, 'Search for exact query(True or False, default=False)', '-e', 'store', bool),
```
'store' to 'store_true' this is a bool option.

saeeddhqan commented on modules/search/zlibrary.py:
```diff
@@ -0,0 +1,105 @@
+
+	for table in raw_data.find_all('table', {'class': 'table_book'}):
+	
+		if output_table == []: #If this statement is not used, table will reset each time and only 50 entries will be there
```
comments should be above the condition. this format for all

saeeddhqan commented on modules/search/zlibrary.py:
```diff
@@ -0,0 +1,105 @@
+			output_table = [ {header.text:[]} for header in table.findAll('th', {'class': 'header'}) ]
+		
+		for row in table.find_all('tr'):
+			for row_data, column in zip(row.find_all('td'), output_table): #Using zip to avoid cartesian product caused by 2 loops
```
comments should be above the loop. this format for all

saeeddhqan commented on modules/search/zlibrary.py:
```diff
@@ -0,0 +1,105 @@
+		
+		for row in table.find_all('tr'):
+			for row_data, column in zip(row.find_all('td'), output_table): #Using zip to avoid cartesian product caused by 2 loops
+				row_data = " ".join(row_data.text.split()) #Using split to clean the data and remove trailing spaces, /n, etc.
```
quote error. comments should be above the variable. this format for all

vikas-kundu committed:
Minor fixes

vikas-kundu commented:
- Added a destructor in util class to reset random agent value to what it was before.
- Fixed space errors
- Fixed comment space errors
- Fixed string formating errors for quotes
- Fixed store_true error
- Fixed framework.error format

z3y50n commented on modules/search/zlibrary.py:
```diff
@@ -0,0 +1,109 @@
+	'sources': ('Zlibrary',),
+	'options': (
+		('query', '', True, 'Query string to search for', '-q', 'store', str),
+		('exact', False, False, 'Search for exact query(True or False, default=False)', '-e', 'store_true', bool),
```
Here you mention that `exact` is a `bool` but in the `examples` on line 36 you input `pdf`

saeeddhqan replied:
bool options should not have a short option. so change -e to --exact

z3y50n commented:
Also I think you shouldn't use a table to display results because it is not very readable in my opinion as it currently is

vikas-kundu committed:
Example improve

vikas-kundu commented:
> Also I think you shouldn't use a table to display results because it is not very readable in my opinion as it currently is

Using output as it is would be more unreadable. So, should i print tables in the format as the Worldcat module does?

saeeddhqan commented on modules/search/zlibrary.py:
```diff
@@ -0,0 +1,109 @@
+		('ext', '', False, 'Extension of the ebook or article(i.e. pdf, epub, txt, rar, mobi)', '-ex', 'store', str),
+		('count', 50, False, 'No. of results to show(default=50)', '-c', 'store', int) # 50 entries per page
+	),
+    'examples': ('zlibrary -q <QUERY> -ex pdf',)
```
-q ... --ext/-e pdf

vikas-kundu replied:
can't do that, exact option is already -e

saeeddhqan replied:
bool options should not have a short option. so change -e to --exact

vikas-kundu replied:
Leaving bool option empty causes parser errors. So should i change its short option to -exact ?

saeeddhqan replied:
So should i change its short option to -exact? I mean it

z3y50n commented:
> > Also I think you shouldn't use a table to display results because it is not very readable in my opinion as it currently is
> 
> Using output as it is would be more unreadable. So, should i print tables in the format as the Worldcat module does?

I guess it would be better that way

vikas-kundu committed:
Changed exact options and extension options

saeeddhqan commented:
Everything is okay?

vikas-kundu commented:
> Everything is okay?

Need to work on the table format as @z3y50n said.

saeeddhqan commented on modules/search/zlibrary.py:
```diff
@@ -0,0 +1,109 @@
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+from bs4 import BeautifulSoup
```
use lxml instead of bs4 because lxml is pretty faster than bs4.

vikas-kundu replied:
This would require changes to the entire util and module codeb. Instead I have used lxml parser of beautiful soup instead of HTML parser in latest commit.

vikas-kundu committed:
Output table imrpovement, Author name digit removal, output queries=limit

vikas-kundu commented:
Did the following changes:

- Output table format changed.
- Improved author name by removing digit in front of it.
- Result contains queries as much as the count option only.
